### ISSUE TITLE
segmentation fault!! No other failed messages.

### ISSUE THREAD

**Issue by fuxuemingzhu** [aksnzhy/xlearn#99]
Thanks to your wonderful works! Here is my failed message.

I don't have other information to check the error point. Could you help me?

I ran the same code for three times and get the same error.

My environment is Python 3.5.2, Ubuntu 16.04. The RAM is 32GB.

HERE is my error messages:

```
[ ACTION     ] Read Problem ...
[------------] First check if the text file has been already converted to binary format.
[------------] Binary file (../data/cfg/ffm_train.txt.bin) found. Skip converting text to binary.
[------------] First check if the text file has been already converted to binary format.
[------------] Binary file (../data/cfg/ffm_valid.txt.bin) found. Skip converting text to binary.
[------------] Number of Feature: 32870706
[------------] Number of Field: 50
[------------] Time cost for reading problem: 38.60 (sec)
[ ACTION     ] Initialize model ...
[1]    26156 segmentation fault (core dumped)  python3 train_xlearn.py
```

**Comment by aksnzhy** [comment]
It seems that your data format is not correct. Can you share apart of your data?

**Comment by fuxuemingzhu** [comment]
Here is head 10 lines of my data. It contains int,str,float type.

I used the [pandas2libFFM](https://www.kaggle.com/mpearmain/pandas2libffm) to convet pandas into libffm format. And I added below line in function ``transform_row_``.

```python
elif col_type.kind == 'f':
    ffm.append('{}:{}:{:.6f}'.format(self.field_index_[col], self.feature_index_[col], val))
```

Can you give me a quite good method to convert pandas DataFrame or raw csv into libffm format? Or is there a function to check whether my libffm format is correct?

```
0 2:82:17597 8:221:51385 11:395:838056 12:411:35 1:452:25 42:486:0 43:491:6 0:1328:810.000000 3:1335:4 4:1336:1 6:91905:1 9:231747:1 10:231751:2 13:231752:1 15:231826:3 16:231830:2 17:231831:1 18:231834:1 20:4319217:1 22:5345589:1 24:5345894:1 26:5346191:1 32:10014454:1 34:16291908:1 36:21832328:1 38:22074239:1 40:22074266:1 44:22074272:1 46:28487251:1 48:32572602:1 5:32869589:0 7:32870470:0 14:32870475:1 41:32870478:1 39:32870482:2 19:32870521:8 21:32870555:2 23:32870567:0 25:32870579:0 27:32870663:12 33:32870670:5 35:32870677:5 37:32870684:0 45:32870691:5 47:32870698:5 49:32870705:0
0 2:82:21359 8:221:361928 11:395:585909 12:411:100 1:452:21 42:486:0 43:491:4 0:1328:94.000000 3:1335:3 4:1336:1 6:91905:1 9:231747:0 10:231751:0 13:231753:1 15:231826:6 16:231830:2 17:231831:1 18:231835:1 20:4319218:1 22:5345589:1 24:5345894:1 26:5346192:1 32:10014455:1 34:16291909:1 36:21832328:1 38:22074240:1 40:22074267:1 44:22074273:1 46:28487252:1 48:32572602:1 5:32869589:0 7:32870470:0 14:32870475:2 41:32870478:1 39:32870482:1 19:32870521:29 21:32870555:2 23:32870567:0 25:32870579:0 27:32870663:0 33:32870670:5 35:32870677:5 37:32870684:0 45:32870691:5 47:32870698:5 49:32870705:0
0 2:82:8203 8:221:47118 11:395:1083112 12:411:95 1:452:218 42:486:0 43:491:6 0:1328:686.000000 3:1335:4 4:1336:1 6:91905:1 9:231747:1 10:231751:2 13:231754:1 15:231826:1 16:231830:1 17:231831:1 18:231836:1 20:4319219:1 22:5345589:1 24:5345894:1 26:5346193:1 32:10014456:1 34:16291910:1 36:21832328:1 38:22074240:1 40:22074266:1 44:22074274:1 46:28487253:1 48:32572602:1 5:32869589:0 7:32870470:0 14:32870475:1 41:32870478:1 39:32870482:1 19:32870521:16 21:32870555:0 23:32870567:0 25:32870579:0 27:32870663:9 33:32870670:5 35:32870677:5 37:32870684:0 45:32870691:5 47:32870698:5 49:32870705:0
0 2:82:133292 8:221:464828 11:395:1334609 12:411:22 1:452:74 42:486:0 43:491:4 0:1328:418.000000 3:1335:4 4:1336:1 6:91906:1 9:231747:0 10:231751:2 13:231755:1 15:231826:2 16:231830:1 17:231831:1 18:231837:1 20:4319220:1 22:5345590:1 24:5345894:1 26:5346192:1 32:10014457:1 34:16291911:1 36:21832329:1 38:22074240:1 40:22074266:1 44:22074275:1 46:28487254:1 48:32572603:1 5:32869589:0 7:32870470:9 14:32870475:2 41:32870478:1 39:32870482:1 19:32870521:31 21:32870555:6 23:32870567:10 25:32870579:0 27:32870663:0 33:32870670:5 35:32870677:5 37:32870684:5 45:32870691:5 47:32870698:5 49:32870705:5
0 2:82:6946 8:221:296367 11:395:455396 12:411:59 1:452:24 42:486:3794 43:491:11 0:1328:839.000000 3:1335:2 4:1336:1 6:91905:1 9:231747:0 10:231751:1 13:231756:1 15:231826:7 16:231830:1 17:231831:1 18:231838:1 20:4319219:1 22:5345589:1 24:5345894:1 26:5346192:1 32:10014458:1 34:16291912:1 36:21832328:1 38:22074241:1 40:22074268:1 44:22074276:1 46:28487255:1 48:32572602:1 5:32869589:0 7:32870470:0 14:32870475:1 41:32870478:1 39:32870482:1 19:32870521:0 21:32870555:0 23:32870567:0 25:32870579:0 27:32870663:0 33:32870670:0 35:32870677:5 37:32870684:0 45:32870691:0 47:32870698:5 49:32870705:0
0 2:82:158679 8:221:643438 11:395:1690612 12:411:60 1:452:4 42:486:0 43:491:4 0:1328:395.000000 3:1335:4 4:1336:1 6:91905:1 9:231747:1 10:231751:0 13:231755:1 15:231826:7 16:231830:1 17:231831:1 18:231839:1 20:4319221:1 22:5345589:1 24:5345894:1 26:5346194:1 32:10014459:1 34:16291913:1 36:21832328:1 38:22074240:1 40:22074266:1 44:22074277:1 46:28487256:1 48:32572602:1 5:32869589:0 7:32870470:0 14:32870475:2 41:32870478:1 39:32870482:1 19:32870521:30 21:32870555:4 23:32870567:0 25:32870579:0 27:32870663:20 33:32870670:5 35:32870677:5 37:32870684:0 45:32870691:5 47:32870698:5 49:32870705:0
0 2:82:1082 8:221:295940 11:395:1391569 12:411:59 1:452:22 42:486:0 43:491:6 0:1328:275.000000 3:1335:1 4:1336:1 6:91905:1 9:231747:1 10:231751:1 13:231757:1 15:231826:7 16:231830:2 17:231831:1 18:231840:1 20:4319219:1 22:5345589:1 24:5345894:1 26:5346195:1 32:10014458:1 34:16291914:1 36:21832328:1 38:22074240:1 40:22074267:1 44:22074278:1 46:28487257:1 48:32572602:1 5:32869589:0 7:32870470:0 14:32870475:3 41:32870478:1 39:32870482:1 19:32870521:4 21:32870555:0 23:32870567:0 25:32870579:0 27:32870663:34 33:32870670:0 35:32870677:5 37:32870684:0 45:32870691:5 47:32870698:5 49:32870705:0
0 2:82:5459 8:221:172796 11:395:1360276 12:411:59 1:452:142 42:486:0 43:491:6 0:1328:346.000000 3:1335:1 4:1336:1 6:91905:1 9:231747:0 10:231751:1 13:231755:1 15:231826:7 16:231830:1 17:231831:1 18:231841:1 20:4319219:1 22:5345589:1 24:5345894:1 26:5346192:1 32:10014460:1 34:16291915:1 36:21832328:1 38:22074242:1 40:22074266:1 44:22074279:1 46:28487258:1 48:32572602:1 5:32869589:0 7:32870470:0 14:32870475:2 41:32870478:1 39:32870482:2 19:32870521:14 21:32870555:0 23:32870567:0 25:32870579:0 27:32870663:0 33:32870670:5 35:32870677:3 37:32870684:0 45:32870691:5 47:32870698:5 49:32870705:0
0 2:82:9106 8:221:662422 11:395:1354071 12:411:79 1:452:21 42:486:0 43:491:4 0:1328:421.000000 3:1335:2 4:1336:1 6:91905:1 9:231747:1 10:231751:1 13:231756:1 15:231826:1 16:231830:1 17:231831:1 18:231842:1 20:4319219:1 22:5345589:1 24:5345894:1 26:5346196:1 32:10014461:1 34:16291916:1 36:21832328:1 38:22074241:1 40:22074268:1 44:22074280:1 46:28487259:1 48:32572602:1 5:32869589:0 7:32870470:0 14:32870475:1 41:32870478:1 39:32870482:1 19:32870521:3 21:32870555:0 23:32870567:0 25:32870579:0 27:32870663:33 33:32870670:2 35:32870677:5 37:32870684:0 45:32870691:5 47:32870698:5 49:32870705:0
0 2:82:388 8:221:209098 11:395:1146648 12:411:35 1:452:27 42:486:113 43:491:9 0:1328:194.000000 3:1335:5 4:1336:1 6:91905:1 9:231747:0 10:231751:2 13:231756:1 15:231826:2 16:231830:2 17:231831:1 18:231838:1 20:4319222:1 22:5345589:1 24:5345894:1 26:5346197:1 32:10014458:1 34:16291917:1 36:21832328:1 38:22074241:1 40:22074268:1 44:22074276:1 46:28487260:1 48:32572602:1 5:32869589:0 7:32870470:0 14:32870475:1 41:32870478:1 39:32870482:1 19:32870521:0 21:32870555:2 23:32870567:0 25:32870579:0 27:32870663:4 33:32870670:0 35:32870677:5 37:32870684:0 45:32870691:0 47:32870698:5 49:32870705:0
```

**Comment by aksnzhy** [comment]
It seems that your model is too large. Can you show me the hyper-parameter? I will check how may the memory usage in your task.

**Comment by fuxuemingzhu** [comment]
The hyper-parameter is:
```python
param = {'task'  : 'binary', 'lr': 0.2,
         'lambda': 0.002, 'metric': 'auc'}
```
The train set is about 4.4GB. The valid set is about 470MB. The test set is about 1.3GB.

Thanks.

**Comment by fuxuemingzhu** [comment]
Sorry for this issues. My fault to make an incorrect libffm format dataset.

The proccess of float type is wrong.

Grateful for your reply and I will close this issue.

**Comment by MrCuri0sity** [comment]
@fuxuemingzhu i wandered what the correct float type , i meet the same problem

**Comment by fuxuemingzhu** [comment]
@MrCuri0sity Is your model also too large? The Number of Feature being about 1 million is fine.

My float type is correct, but this problem occured when the number of features is large.

The solution is using hash trick.

```python
def hashstr(str, nr_bins):
    return int(hashlib.md5(str.encode('utf8')).hexdigest(), 16)%(nr_bins-1)+1
```

You can get more information from [3 Idiots' Approach for Display Advertising Challenge](https://github.com/guestwalk/kaggle-2014-criteo/blob/master/converters/common.py).

**Comment by MrCuri0sity** [comment]
@fuxuemingzhu after i reduce my features, it works. many thanks!

**Comment by xxllp** [comment]
为啥我特征不多也有这个报错，这个很耗内存吗
